Optional OpenLR Encoded Path Edges in API Response

mookerji opened the pull request:
Adds support for a new request option 'linear_references=true|false', which adds an array of OpenLR references to the response body for `route` and `trace_route` Valhalla HTTP endpoints.

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -1,11 +1,18 @@
+#include <math.h>
```
should be `<cmath>`

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -283,23 +206,85 @@ json::MapPtr geojson_shape(const std::vector<PointLL> shape) {
 }
 
+// Maps Valhalla edge and road class to an OpenLR form-of-way.
+FormOfWay RoadClassToFOW(const TripLeg::Node& node) {
```
Could we add `FoW` as a member to the protobuf trip edge and set it in triplegbuilder. That way all downstream uses of it wil be supported? There is a function in that source file called `AddTripEdge` i think it could be added there.

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -283,23 +206,85 @@ json::MapPtr geojson_shape(const std::vector<PointLL> shape) {
+  }
+}
+
+// TODO(mookerji): Move this shared header file somewhere and remove also from tests/openlr.cc
```
yeah i feel like we may as well just move this to the openlr header itself and provide an option to the `toBinary(bool b64_encode = false)` method to optionally b64 encode the results

kevinkreiser replied:
same with the constructor that takes the binary reference. should become `LinearReference(const std::string& reference, bool b64_encoded = false)`

mookerji replied:
Done

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -283,23 +206,85 @@ json::MapPtr geojson_shape(const std::vector<PointLL> shape) {
+}
+
+// TODO(mookerji): comment
+std::string openlr_shape(const valhalla::TripRoute& route) {
```
the first and last edges of a leg may (and probably will) have a point that is not at a graph node (intersection) at the beginning or end of the edge respectively. i think the canonical way to handle such openlrs is to actually use the point at the preceding or following intersection respectively and then set positive offset (poff) or negative offset (noff) respectively. i dont think we need to handle that in the first go round here. but in the future it might make sense, in triplegbuilder, to add some information to the leg abotu the interection the edge starts/ends on as opposed to the point the route starts/ends on.

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -283,23 +206,85 @@ json::MapPtr geojson_shape(const std::vector<PointLL> shape) {
+        bearing_deg = location.Heading(next_location);
+        distance_meters = location.Distance(next_location);
+      }
+      // TODO(continued): should the start node for a new leg point to the lrps of the last node of
```
yeah i think that because you smashing all the legs together it would make more sense to only add the last lrp of a leg if it is the last leg (and also only then flip the bearing). basically add a break to the top of the nodes for loop that is like

```c++
if(&leg == &route.legs().back())
  break;
```

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -283,23 +206,85 @@ json::MapPtr geojson_shape(const std::vector<PointLL> shape) {
+                        lrps.empty() ? nullptr : &lrps.back(), distance_meters);
+    }
+  }
+  // TODO(mookerji): implement lfrcnp
```
should be just `lowest_frc = std::min(lowest_frc, frc)` yeah?

mookerji replied:
Yup

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -283,23 +206,85 @@ json::MapPtr geojson_shape(const std::vector<PointLL> shape) {
+    }
+  }
+  // TODO(mookerji): implement lfrcnp
+  return encode64(OpenLR::LineLocation{lrps, 0, 0}.toBinary());
```
i originally assumed that we'd do a list of openlrs, one for each edge to ease the use of this on the client side application. however doing them all as one large linear reference is quite a compactified representation in comparison. it would be interesting to see how much difference in size? its kind of nice to do the openlrs as exact edges in the graph and i feel like by breaking them up its more explicitly so. i havent thought about it much though, both have pros and cons. whats your take?

mookerji replied:
Updated, also see follow up comment in the thread below.

mookerji commented:
Per off-Github discussion, moved this from the geometry field to a new field in the map matching response that returns a single OpenLR descriptor _per edge_. Only a small subset of clients need such descriptors, so I've also added an optional request flag; otherwise, the default response size will increase quite a bit. 

```
$ curl --data '{"shape":[{"lat":"37.7826603","lon":"-122.4708554","type":"break","radius":15,"search_cutoff":21000000,"preferred_side":"either"},{"lat":"37.7826734","lon":"-122.4706716","type":"break","radius":15,"search_cutoff":21000000,"preferred_side":"either"},{"lat":"37.7826703","lon":"-122.4705267","type":"break","radius":15,"search_cutoff":21000000,"preferred_side":"either"}],"costing":"auto","date_time":{"type":0},"shape_match":"map_snap","format":"osrm","shape_format":"geojson","filters":{},"include_linear_reference": true, "trace_options":{"interpolation_distance":0},"force_traffic":false,"directions_type":"instructions","units":"miles"}' localhost:8004/trace_route
{
  "matchings": [
    {
      "weight_name": "Valhalla default",
      "weight": 2.608,
      "legs": [
      ....        
      ],
      "duration": 2.608,
      "distance": 28.979,
      "geometry": {
        "coordinates": [
          [
            -122.470856,
            37.782661
          ],
          [
            -122.470673,
            37.782669
          ],
          [
            -122.470528,
            37.782677
          ]
        ],
        "type": "LineString"
      },
      "linear_reference": [
        "C6jo3BreID/nAAASAAE/Fw==",
        "C6jo3BreID/gAAAAAAA/EA==",
        "C6jo5RreID/nAAANAAI/Fw==",
        "C6jo5RreID/gAP//AAE/EA=="
      ],
      "confidence": 1
    }
  ],
  "tracepoints": [
    {
      "matchings_index": 0,
      "name": "",
      "waypoint_index": 0,
      "alternatives_count": 0,
      "distance": 0,
      "location": [
        -122.470856,
        37.782661
      ]
    },
    {
      "matchings_index": 0,
      "name": "",
      "waypoint_index": 1,
      "alternatives_count": 0,
      "distance": 0.425,
      "location": [
        -122.470673,
        37.782669
      ]
    },
    {
      "matchings_index": 0,
      "name": "",
      "waypoint_index": 2,
      "alternatives_count": 0,
      "distance": 0.85,
      "location": [
        -122.470528,
        37.782677
      ]
    }
  ],
  "code": "Ok"
}
```

mookerji commented on test/util_midgard.cc:
```diff
@@ -459,6 +459,27 @@ TEST(UtilMidgard, TrimShapeEmpty) {
   ASSERT_EQ(shape.size(), 0);
 }
 
+// Test cases from: https://tools.ietf.org/html/rfc4648#section-10
```
@kevinkreiser I moved the base64 encoding stuff into a header and added some test cases from the base64 encoding RFC and ran into some null termination issues. I expected roundtrip equality here, but It looks like we're adding some extra bytes here? 

```
/Users/buromookerji/projects/mapbox/valhalla/test/util_midgard.cc:478: Failure
Expected equality of these values:
  decode64(encoded)
    Which is: "f\0\0"
  decoded
    Which is: "f"
/Users/buromookerji/projects/mapbox/valhalla/test/util_midgard.cc:479: Failure
Expected equality of these values:
  decode64(encode64(decoded))
    Which is: "f\0\0"
  decoded
    Which is: "f"
/Users/buromookerji/projects/mapbox/valhalla/test/util_midgard.cc:478: Failure
Expected equality of these values:
  decode64(encoded)
    Which is: "fo\0"
  decoded
    Which is: "fo"
/Users/buromookerji/projects/mapbox/valhalla/test/util_midgard.cc:479: Failure
Expected equality of these values:
  decode64(encode64(decoded))
    Which is: "fo\0"
  decoded
    Which is: "fo"
/Users/buromookerji/projects/mapbox/valhalla/test/util_midgard.cc:478: Failure
Expected equality of these values:
  decode64(encoded)
    Which is: "foob\0\0"
  decoded
    Which is: "foob"
/Users/buromookerji/projects/mapbox/valhalla/test/util_midgard.cc:479: Failure
Expected equality of these values:
  decode64(encode64(decoded))
    Which is: "foob\0\0"
  decoded
    Which is: "foob"
/Users/buromookerji/projects/mapbox/valhalla/test/util_midgard.cc:478: Failure
Expected equality of these values:
  decode64(encoded)
    Which is: "fooba\0"
  decoded
    Which is: "fooba"
/Users/buromookerji/projects/mapbox/valhalla/test/util_midgard.cc:479: Failure
Expected equality of these values:
  decode64(encode64(decoded))
    Which is: "fooba\0"
  decoded
    Which is: "fooba"
```

mookerji replied:
Fixed

kevinkreiser commented on docs/api/map-matching/api-reference.md:
```diff
@@ -141,6 +141,15 @@ matched.distance_along_edge
+
+`include_linear_references` is an optional boolean input parameter. When `true`, the successful
+`trace_route` response will include a key `linear_references`. Its value is an array of
+base64-encoded [OpenLR location references][poenlr], one for each graph edge of the road network
```
typo `poenlr`

mookerji replied:
Fixed

kevinkreiser commented on proto/options.proto:
```diff
@@ -189,4 +189,5 @@ message Options {
   // 42 is reserved
   optional uint32 height_precision = 43 [default = 0];                    // Number of digits precision for heights returned
   optional bool roundabout_exits = 44 [default = true];                   // Whether to announce roundabout exit maneuvers
+  optional bool include_linear_references = 45;                           // Include linear references for graph edges returned in certain responses.
```
i think we can drop the `include_` part of the name to be in line with other parameters like `roundabout_exits` and `guidance_views`

mookerji replied:
Great idea.

kevinkreiser commented on src/midgard/util.cc:
```diff
@@ -697,5 +701,18 @@ polygon_t to_boundary(const std::unordered_set<uint32_t>& region, const Tiles<Po
+  using namespace boost::archive::iterators;
+  using It = base64_from_binary<transform_width<std::string::const_iterator, 6, 8>>;
+  auto tmp = std::string(It(std::begin(val)), It(std::end(val)));
+  return tmp.append((3 - val.size() % 3) % 3, '=');
```
@mookerji this is where the padding comes from

kevinkreiser replied:
yeah im pretty sure this is where the padding comes from. i think at some point some use-case must have needed to pad to the nearest multiple of 3 bytes but im failing to remember where/why/what...

kevinkreiser replied:
https://stackoverflow.com/questions/4080988/why-does-base64-encoding-require-padding-if-the-input-length-is-not-divisible-by/18518605#18518605

kevinkreiser replied:
it seems like we need to modify the decode function to ignore the padding

kevinkreiser replied:
which i think can be done by removing null characters from the end?

mookerji replied:
Fixed

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -283,23 +195,36 @@ json::MapPtr geojson_shape(const std::vector<PointLL> shape) {
 
+// Return a JSON array of OpenLR 1.5 line location references for each edge of a map matching
+// result. For the time being, result is only non-empty for auto costing requests.
+json::ArrayPtr route_references(const valhalla::TripRoute& route, const valhalla::Options& options) {
```
can we add this the valhalla serializer as well

mookerji replied:
Ah, I knew you were going to say that.

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -1378,12 +1296,16 @@ std::string serialize(valhalla::Api& api) {
 
-    // TODO: phase 1, just hardcode score. phase 2: do real implementation
-    if (options.action() == valhalla::Options::trace_route)
+    if (options.action() == valhalla::Options::trace_route) {
```
this should not be limited to map matching.

mookerji replied:
Yep

kevinkreiser commented on test/mapmatch.cc:
```diff
@@ -1428,10 +1421,50 @@ TEST(Mapmatch, test_breakage_distance_error_handling) {
   FAIL() << "Expected trace_route breakage distance exceed exception was not found";
 }
+
+// Spot check OpenLR linear references when asked for
```
imho these would be better off in a new gurka test designed specifically to verify these for both map matching and regular routes

mookerji replied:
Yup

mookerji commented:
This isochrone [unit test failure][test] is deterministic, but seems out-of-scope of the current change. Checking this out right now, but any other feedback you have @kevinkreiser would be great.

[test]: https://circleci.com/gh/valhalla/valhalla/25559?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

kevinkreiser commented on src/midgard/openlr.cc:
```diff
@@ -0,0 +1,97 @@
+#include <cmath>
+#include <string>
```
We should move this to the baldr namespace since it's not just geometry stuff but really a path on the road network with road network attribution etc.

mookerji replied:
Done

kevinkreiser commented on test/predictedspeeds.cc:
```diff
@@ -63,7 +52,7 @@ TEST(PredicteSpeeds, test_decoding) {
 
   // Decode the base64 string and cast the data to a raw string of signed bytes
   auto decoded_str = decode64(encoded_speed_string);
-  EXPECT_EQ(decoded_str.size(), 402);
```
This original and new value don't make sense to me. There are 200 coefficients each should be 2 bytes wide so we should have 400 here... Are the decoded values different, I noticed the test below doesn't test equality, maybe it was doing this because of the decoding bug you fixed

mookerji replied:
Think I figured this out, noted here: https://github.com/valhalla/valhalla/pull/2452

mookerji commented:
Factored out the base64 stuff into: https://github.com/valhalla/valhalla/pull/2452/files. Once we get that merged, we can rebase this OpenLR PR on master and review again.
